### ISSUE TITLE
Re-enable & fix unit tests in test_avrotools.py for schemas_e6d6074

### ISSUE THREAD

**Issue by macieksmuga** [ga4gh/server#357]
PR #353 cleanup

**Comment by dcolligan** [comment]
This is failing because the schemas are not getting written out correctly.  We could either fix the schema ourselves or add some more exceptions in `ga4gh.avrotools`.  (While we are evaluating this issue we should see if the exception in `ga4gh.avrotools` is still necessary.)

Errors include:
```
(for the protocol.Individual class... a similar error happens for the protocol.Sample class)
ERROR: testGeneratedObjects (tests.unit.test_avrotools.TestValidationTool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/server/tests/unit/test_avrotools.py", line 47, in testGeneratedObjects
    class_, jsonDict)
  File "/Users/dannyc/ga4gh/server/ga4gh/avrotools.py", line 55, in getInvalidFields
    fieldSchema = cls._createFieldSchema(schema, fieldName)
  File "/Users/dannyc/ga4gh/server/ga4gh/avrotools.py", line 97, in _createFieldSchema
    fieldSchema = avro.schema.parse(fieldSchemaSource)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 784, in parse
    return make_avsc_object(json_data, names)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 740, in make_avsc_object
    return RecordSchema(name, namespace, fields, names, type, doc, other_props)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 661, in __init__
    field_objects = RecordSchema.make_field_objects(fields, names)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 627, in make_field_objects
    other_props)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 342, in __init__
    raise SchemaParseException(fail_msg)
SchemaParseException: Type property "[u'null', u'org.ga4gh.models.OntologyTerm']" not a valid Avro schema: Union item must be a valid Avro schema: Could not make an Avro Schema object from org.ga4gh.models.OntologyTerm.
```
and
```
(for class protocol.ReadGroupSet)
ERROR: testGeneratedObjects (tests.unit.test_avrotools.TestValidationTool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/server/tests/unit/test_avrotools.py", line 49, in testGeneratedObjects
    class_, jsonDict)
  File "/Users/dannyc/ga4gh/server/ga4gh/avrotools.py", line 55, in getInvalidFields
    fieldSchema = cls._createFieldSchema(schema, fieldName)
  File "/Users/dannyc/ga4gh/server/ga4gh/avrotools.py", line 97, in _createFieldSchema
    fieldSchema = avro.schema.parse(fieldSchemaSource)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 784, in parse
    return make_avsc_object(json_data, names)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 740, in make_avsc_object
    return RecordSchema(name, namespace, fields, names, type, doc, other_props)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 661, in __init__
    field_objects = RecordSchema.make_field_objects(fields, names)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 627, in make_field_objects
    other_props)
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/avro/schema.py", line 342, in __init__
    raise SchemaParseException(fail_msg)
SchemaParseException: Type property "{u'items': {u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'ReadGroup', u'fields': [{u'doc': u'', u'type': u'string', u'name': u'id'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'datasetId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'name'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'description'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'sampleId'}, {u'doc': u'', u'type': [u'null', {u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'Experiment', u'fields': [{u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'libraryId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'platformUnit'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'sequencingCenter'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'instrumentModel'}]}], u'name': u'experiment'}, {u'default': None, u'doc': u'', u'type': [u'null', u'int'], u'name': u'predictedInsertSize'}, {u'default': None, u'doc': u'', u'type': [u'null', u'long'], u'name': u'created'}, {u'default': None, u'doc': u'', u'type': [u'null', u'long'], u'name': u'updated'}, {u'default': None, u'doc': u'', u'type': [u'null', u'org.ga4gh.models.ReadStats'], u'name': u'stats'}, {u'default': [], u'doc': u'', u'type': {u'items': {u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'Program', u'fields': [{u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'commandLine'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'id'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'name'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'prevProgramId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'version'}]}, u'type': u'array'}, u'name': u'programs'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'referenceSetId'}, {u'default': {}, u'doc': u'', u'type': {u'values': {u'items': u'string', u'type': u'array'}, u'type': u'map'}, u'name': u'info'}]}, u'type': u'array'}" not a valid Avro schema: Items schema ({u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'ReadGroup', u'fields': [{u'doc': u'', u'type': u'string', u'name': u'id'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'datasetId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'name'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'description'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'sampleId'}, {u'doc': u'', u'type': [u'null', {u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'Experiment', u'fields': [{u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'libraryId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'platformUnit'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'sequencingCenter'}, {u'doc': u'', u'type': [u'null', u'string'], u'name': u'instrumentModel'}]}], u'name': u'experiment'}, {u'default': None, u'doc': u'', u'type': [u'null', u'int'], u'name': u'predictedInsertSize'}, {u'default': None, u'doc': u'', u'type': [u'null', u'long'], u'name': u'created'}, {u'default': None, u'doc': u'', u'type': [u'null', u'long'], u'name': u'updated'}, {u'default': None, u'doc': u'', u'type': [u'null', u'org.ga4gh.models.ReadStats'], u'name': u'stats'}, {u'default': [], u'doc': u'', u'type': {u'items': {u'namespace': u'org.ga4gh.models', u'type': u'record', u'name': u'Program', u'fields': [{u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'commandLine'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'id'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'name'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'prevProgramId'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'version'}]}, u'type': u'array'}, u'name': u'programs'}, {u'default': None, u'doc': u'', u'type': [u'null', u'string'], u'name': u'referenceSetId'}, {u'default': {}, u'doc': u'', u'type': {u'values': {u'items': u'string', u'type': u'array'}, u'type': u'map'}, u'name': u'info'}]}) not a valid Avro schema: Type property "[u'null', u'org.ga4gh.models.ReadStats']" not a valid Avro schema: Union item must be a valid Avro schema: Could not make an Avro Schema object from org.ga4gh.models.ReadStats. (known names: [u'org.ga4gh.models.ReadGroupSet', u'org.ga4gh.models.Experiment', u'org.ga4gh.models.ReadGroup'])
```
and
```
(for class SearchExperimentsResponse)
FAIL: testGeneratedObjects (tests.unit.test_avrotools.TestValidationTool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/server/tests/unit/test_avrotools.py", line 50, in testGeneratedObjects
    self.assertEqual(returnValue, {})
AssertionError: {u'experiments': [{u'sequencingCenter': None, u'libraryId': None, u'instrumentMo [truncated]... != {}
Diff is 1997 characters long. Set self.maxDiff to None to see it.
```

**Comment by dcolligan** [comment]
Ok, with the new recursive avrotools I just wrote, I've narrowed down the error fields.  The only one failing now is `SearchExperimentsResponse`.  It's failing because the `id` and `info` fields need to be non-null.
```
+ {}
- {u'experiments': [{u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None},
-                   {u'id': None, u'info': None}]}
```

**Comment by dcolligan** [comment]
I'm not sure if this is a problem with our `tests.utils.InstanceGenerator` or the schemas, but the schema for `SearchExperimentsResponse` seems inconsistent with the `Experiment` types it is supposedly embedding.  SER's experiments have these fields: `[u'id', u'name', u'description', u'created', u'updated', u'runDate', u'molecule', u'strategy', u'selection', u'library', u'libraryLayout', u'instrumentModel', u'instrumentDataFile', u'sequencingCenter', u'platformUnit', u'info']` whereas Experiment's schema defines these fields: `['instrumentModel', 'libraryId', 'platformUnit', 'sequencingCenter']` ... @macieksmuga  your thoughts?

**Comment by dcolligan** [comment]
Fixed in #515
